Dont segfault if btree range is not in order

bvinc opened the pull request:
This is a first attempt to fix issue #33197.  The issue is that the BTree iterator uses next_unchecked for fast iteration, but it can be tricked into running off the end of the tree and segfaulting if range is called with a maximum that is less than the minimum.

Since a user defined Ord should not determine the safety of BTreeMap, and we still want fast iteration, I've implemented the idea of @gereeter and walk the tree simultaneously searching for both keys to make sure that if our keys diverge, the min key is to the left of our max key.  I currently panic if that is not the case.

Open questions:

1.  Do we want to panic in this error case or do we want to return an empty iterator?  The drain API panics if the range is bad, but drain is given a range of index values, while this is a generic key type.  Panicking is brittle and returning an empty iterator is probably the most flexible and matches what people would want it to do... but artificially returning a BTreeMap::Range with start==end seems like a pretty weird and unnatural thing to do, although it's doable since those fields are not accessible.

The same question for other weird cases:
2.  (Included(101), Excluded(100)) on a map that contains [1,2,3].  Both BTree edges end up on the same part of the map, but comparing the keys shows the range is backwards.
3.  (Excluded(5), Excluded(5)).  The keys are equal but BTree edges end up backwards if the map contains 5.
4.  (Included(5), Excluded(5)).  Should naturally produce an empty iterator, right?

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @BurntSushi (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

BurntSushi commented on src/libcollections/btree/map.rs:
```diff
@@ -689,61 +689,69 @@ impl<K: Ord, V> BTreeMap<K, V> {
+
+            min_edge = match (min_found, min) {
+                (false, Included(key)) => {
+                    match search::search_linear(&min_node, key) {
```
Why change the `search_tree` calls to `search_linear`? (FYI, I'm not familiar with the btree impl.)

bvinc replied:
search_tree searches the entire tree and returns a handle anywhere in the tree.  I need to search a single node at a time and be able to compare 2 searches.  You might think that I want search_node, but that returns a handle.  I need to be able to compare 2 edges of the same node, which I can't.  search_linear just gives me an index value into the node, which is exactly what I need.

BurntSushi commented:
cc @rust-lang/libs 

Someone more familiar with the btree impl should review this. cc @gereeter 

With respect to the panic, I think that's consistent behavior with other collections. Either way, I think this behavior should be documented.

alexcrichton commented:
also cc @arthurprs  and @apasel422

arthurprs commented:
This is very smart :+1: 

I wonder if it makes sense to have a `fn nodes_diverge(n1, n2)` in node.rs that take 2 handles and returns if `n1.node == n2.node && n1.idx != n2.idx`, I didn't really try it but I guess it can really simplify the range function.

alexcrichton commented:
@bors: delegate=arthurprs

bors commented:
:v: @arthurprs can now approve this pull request

bvinc commented:
I've made some small cleanup changes, implemented range_mut as well, and changed the implementation so that it should never panic, but only return an empty iterator.

@arthurprs That would only get rid of one variable, right?  Personally, it doesn't seem worth it, and it seems to obfuscate the way I only care about comparing them until the point that they diverge, after which I don't care anymore.

bvinc commented:
I'd really like to steer the discussion towards the strange cases of range(..) and what the behavior should be.  I've read all of the rfcs and proposals and documentation that I can find, but no one seems to address these cases.  If you know of where it's documented, let me know.

If it's not currently documented, who makes these decisions?  It might seem like an extremely small detail, but changing behavior later would be a breaking change.

The way I see it, there are 8 interesting cases to think about, and 3 basic ways to handle them:

## Weird cases
1. range(Included(A), Included(A))
2. range(Included(A), Excluded(A))
3. range(Excluded(A), Included(A))
4. range(Excluded(A), Excluded(A))

where B > A:
5. range(Included(B), Included(A))
6. range(Included(B), Excluded(A))
7. range(Excluded(B), Included(A))
8. range(Excluded(B), Excluded(A))

## Option 1
Return an empty iterator on cases 2-3.  Immediately panic on cases 4-8.  During iteration of the tree, if the 2 tree searches "cross directions" in the middle of searching, we know for sure that Ord is ill-defined, and the function will panic.

### Pros:
- Will catch errors early
### Cons:
- Might cause panics in programs where an empty iterator would have been fine
- Must oddly document case #4 as a case that panics

## Option 2
Return an empty iterator on cases 2-4.  Immediately panic on cases 5-8.  Case 4 would be a strange case where in order to prevent a seg fault, one side of the Range might not end up where you would expect in order to force an empty iterator.  Eg: the range.back handle might be on the right side of A, even though A was excluded.

### Pros:
- Can easily document that it panics when min_key > max_key
### Cons:
- Might cause panics in programs where an empty iterator would have been fine
- Case #4 will be a weird special case

## Option 3
Return an empty iterator on cases 2-8, never panic.  Declare that the range returned will always have range.back equal or to the right of range.front.  This will naturally cause 2-8 to all be empty iterators.

### Pros:
- Never panics
- Matches the intuition that you're getting every key where key > min_key && key < max_key. If that's nothing, that's fine.
### Cons:
- Might not catch programmer errors early

gereeter commented on src/libcollections/btree/map.rs:
```diff
@@ -687,63 +687,59 @@ impl<K: Ord, V> BTreeMap<K, V> {
+
+            if !diverged {
+                // Don't allow max_edge to go to the left of min_edge
+                if max_edge < min_edge { max_edge = min_edge; }
```
If we do go this route of never panicking, this also could be an optimization - just search for `max_edge` starting from `min_edge`. This would never produced "crossed nodes" by construction, and would require fewer comparisons. I don't know whether this is worth it though. Additionally, it can be done later.

gereeter commented on src/libcollections/btree/map.rs:
```diff
@@ -777,69 +773,65 @@ impl<K: Ord, V> BTreeMap<K, V> {
-                    }
-                    GoDown(bottom) => bottom,
-                }
+        loop {
```
This code is identical to the `range` case and fairly complex, so I think it would be useful to pull this logic out into a function that is generic over mutability (taking two aliasing root pointers and returning two leaf handles).

gereeter commented on src/libcollections/btree/map.rs:
```diff
@@ -687,63 +687,59 @@ impl<K: Ord, V> BTreeMap<K, V> {
-                    GoDown(bottom) => bottom,
-                }
+        loop {
+            let min_edge = match (min_found, range.start()) {
```
I'm not sure that there is an at-all clean way to do this, but I'm curious whether it would be more performant to check `Unbounded` vs. `Included` vs. `Excluded` before the loop. I expect it doesn't matter, though - it would just be removing some predictable branches, I think.

gereeter commented on src/libcollections/btree/map.rs:
```diff
@@ -777,69 +773,65 @@ impl<K: Ord, V> BTreeMap<K, V> {
                issue = "27787")]
     pub fn range_mut<T: ?Sized, R>(&mut self, range: R) -> RangeMut<K, V>
-        where T: Ord, K: Borrow<T>, R: RangeArgument<T>
+       where T: Ord, K: Borrow<T>, R: RangeArgument<T>
```
Nit: this indentation change should probably not be here.

aturon commented:
cc @rust-lang/libs about the behavioral options laid out in https://github.com/rust-lang/rust/pull/39457#issuecomment-277901686.

Analogous APIs for slices, such as subslicing, tend to panic on nonsensical ranges, which I think corresponds most closely to Option 1. I think this is the right convention to follow, because such ranges are usually produced by a bug.

arthurprs commented:
I'm not fond of panicking in recoverable errors but Option 1 has precedents from the slice api, so that's probably the way to go.

alexcrichton commented:
I agree with @aturon that erring on the side of more panics here seems good. That should be the conservative change here and I think we've got room to return empty iterators in the future if we really need to?

bvinc commented:
I've changed the code to implement option #1 and added unit tests for those cases.  I took the advice of @gereeter and pulled out the range search logic into a common function used by range and range_mut.

arthurprs commented:
@bvinc could you add some docstrings explaining the panic behavior?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit fb91047 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton fb9104768c0991b935e4b5cbc67180e49d425f2c -->

bors commented:
:hourglass: Testing commit fb91047 with merge ea451ec...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1949)

alexcrichton commented:
@bors: retry

* spurious windows failure #38620

On Tue, Feb 14, 2017 at 8:01 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-appveyor
> <https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1949>
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39457#issuecomment-279714466>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95NxubWXK0gk-yoiytj5A6Bc_vzKRks5rcbPKgaJpZM4L0zQe>
> .
>

bors commented:
:hourglass: Testing commit fb91047 with merge 7edcd67...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/201763727)

alexcrichton commented:
@bors: retry

* spurious network failure w/ crates.io

On Wed, Feb 15, 2017 at 1:30 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-travis
> <https://travis-ci.org/rust-lang/rust/builds/201763727>
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39457#issuecomment-279937025>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95Ia5UeoiZyfzCrHO3J7ArUTzCldeks5rcql_gaJpZM4L0zQe>
> .
>

bors commented:
:hourglass: Testing commit fb91047 with merge 4c4925f...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/201950532)

alexcrichton commented:
@bors: retry

* osx linker segfault #38878

On Wed, Feb 15, 2017 at 1:06 PM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-travis
> <https://travis-ci.org/rust-lang/rust/builds/201950532>
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39457#issuecomment-280106396>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95E8GboearsXQq-1nJWu3U2h8jxDjks5rc0yfgaJpZM4L0zQe>
> .
>

bors commented:
:hourglass: Testing commit fb91047 with merge 0d54a22...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/201984747)

alexcrichton commented:
@bors: retry

* sccache wedged #39003

On Wed, Feb 15, 2017 at 2:06 PM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-travis
> <https://travis-ci.org/rust-lang/rust/builds/201984747>
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/39457#issuecomment-280123312>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95C5GCcWO2QxZqu7w1-gf6XStsUTHks5rc1qwgaJpZM4L0zQe>
> .
>

bors commented:
:hourglass: Testing commit fb91047 with merge 4d6019d...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1969), [status-travis](https://travis-ci.org/rust-lang/rust/builds/202041307)
Approved by: alexcrichton
Pushing 4d6019d07a942f5041e3d81f974f34515b024d0a to master...
